File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func diameterOfBinaryTree(root *TreeNode) int {
    if root == nil {
        return 0
    }
    var diameter float64    
    helper(root, &diameter)
    return int(diameter-1)
}

func helper(root *TreeNode, diameter *float64) float64 {
    if root == nil {
        return 0
    }
    left := helper(root.Left, diameter)
    right := helper(root.Right, diameter)
    *diameter = math.Max(*diameter, left+right+1)
    return math.Max(left, right) + 1
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

    /**
     * @param TreeNode $root
     * @return Integer
     */
    function diameterOfBinaryTree($root) {
        $ans = 0;
        $this->helper($root, $ans);
        return $ans;
    }
    
    function helper($root, &$ans){
        if(!$root){
            return 0;
        }
        
        $left = $this->helper($root->left, $ans);
        $right = $this->helper($root->right, $ans);
        $ans = max($ans, $left + $right);
        
        return max($left, $right) + 1;
    }
}